File: 2018summer/07.12/solutions_2015/legacycode/submissions/accepted/tobi.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <map>

using namespace std;
#define FOR(i,a,b) for (int i = (a); i < (b); i++)

int N, from, to, numP, r, dp[1000];
vector<int> adjList[1000];
map<string,int> nameToInt;
bool isProgram[1000], visited[1000];

int getName(string s) {
	if (nameToInt.find(s) == nameToInt.end()) {
		nameToInt[s] = nameToInt.size() - 1;
		isProgram[nameToInt[s]] = (s.length() >= 10 && s.rfind("::PROGRAM") == s.size() - 9);
	}
	return nameToInt[s];
}

bool dfs(int node) {
	if (visited[node]) return false;
	visited[node] = true;
	for (int next : adjList[node]) if (dfs(next)) return true;
	return false;
}

int main() {
	cin >> N;
	FOR(i, 0, N) {
		string fromS, toS;
		cin >> fromS >> numP;
		from = getName(fromS);
		FOR(j, 0, numP) {
			cin >> toS;
			to = getName(toS);
			adjList[to].push_back(from);
		}
	}
	FOR(i, 0, N) if (isProgram[i]) dfs(i);
	FOR(i, 0, N) if (visited[i]) r++;
	cout << N-r << endl;
	
	return 0;
}

File: 2018summer/07.12/solutions_2015/legacycode/submissions/accepted/tobi.java
import java.util.*;

public class tobi {
	public static int N, from, to, numP, r, dp[] = new int[1000];
	public static ArrayList<Integer> adjList[] = new ArrayList[1000];
	public static HashMap<String, Integer> nameToInt = new HashMap<>();
	public static boolean isProgram[] = new boolean[1000], visited[] = new boolean[1000];

	public static int getName(String s) {
		if (nameToInt.get(s) == null) {
			nameToInt.put(s, nameToInt.size());
			isProgram[nameToInt.get(s)] = s.endsWith("::PROGRAM");
		}
		return nameToInt.get(s);
	}

	public static boolean dfs(int node) {
		if (visited[node]) return false;
		visited[node] = true;
		for (int next : adjList[node]) if (dfs(next)) return true;
		return false;
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		for (int i = 0; i < N; i++) adjList[i] = new ArrayList<>();
		for (int i = 0; i < N; i++) {
			from = getName(sc.next());
			numP = sc.nextInt();
			for (int j = 0; j < numP; j++) {
				to = getName(sc.next());
				adjList[to].add(from);
			}
		}
		for (int i = 0; i < N; i++) if (isProgram[i]) dfs(i);
		for (int i = 0; i < N; i++) if (visited[i]) r++;
		System.out.println(N-r);
	}
}
